abi/decode.go

```diff
@@ -260,19 +275,25 @@ func decodeBool(data []byte) (interface{}, error) {
 	}
 }
 
-func readOffset(data []byte, len int) (int, error) {
+func readOffset(data []byte, length int) (int, error) {
+	if len(data) < 32 {
+		return 0, fmt.Errorf("insufficient length of data")
+	}
 	offsetBig := big.NewInt(0).SetBytes(data[0:32])
 	if offsetBig.BitLen() > 63 {
 		return 0, fmt.Errorf("offset larger than int64: %v", offsetBig.Int64())
 	}
 	offset := int(offsetBig.Int64())
-	if offset > len {
-		return 0, fmt.Errorf("offset insufficient %v require %v", len, offset)
+	if offset > length {
```

Comment:
```suggestion
	if len(data) < 32+length {
		return 0, fmt.Errorf("length insufficient %v require %v", len(data), 32+length)
	}
```
At this point we should check already if the length we decode can be accessed on the `input`.